### DISCUSSION TITLE
Different port impedance yields different incident pulse amplitudes?

### DISCUSSION THREAD

**Discussion by 0xCoto** [thliebig/openEMS-Project#87] [Q&A Usage]
Using Python's `AddLumpedPort()`, I created 2 ports with a different impedance and exciting them (sequentially, not simultaneously) appears to give different amplitudes. Tested with a simple microstrip thru line:

### Port 1 (50 Ω):
<img width="936" alt="Screenshot 2023-06-17 at 04 15 41" src="https://github.com/thliebig/openEMS-Project/assets/25392776/399b99b3-ce1c-4c7a-8089-7e0f54ec1472">

### Port 2 (350 Ω):
<img width="948" alt="Screenshot 2023-06-17 at 04 15 13" src="https://github.com/thliebig/openEMS-Project/assets/25392776/4b6a1640-db0b-4a67-8ab0-acf294fcc498">

![image](https://github.com/thliebig/openEMS-Project/assets/25392776/c61d7125-085b-4557-b5f3-c55bb79d28a5)

For some reason, the i2 (input at port 2) appears to carry more energy than i1. Why could that be? This obviously gives incorrect S-Parameter results etc., as the `transmitted/incident` ratio is "biased":

<img width="1114" alt="Screenshot 2023-06-17 at 04 19 48" src="https://github.com/thliebig/openEMS-Project/assets/25392776/1f7302fe-7748-4b41-acad-8dab65d47c71">

**Comment by thliebig** [comment]
I think a lot gets mixed up here very much... Firstly if the MSL is ~50 Ohms, one port will be matched reasonable while the other is not at all. Secondly if you have "mixed" impedance S-Parameter you have to be much more careful how to calculate them, just using voltage ratios is not it anymore... Furthermore you talk about i2 (a current?) carrying more energy??? That is a complete mix-up of units...

**Reply by 0xCoto** [reply]
> Furthermore you talk about i2 (a current?) carrying more energy???

Sorry for the confusion - as I mentioned `i2` is the **i**nput time-domain signal stimulated into port 2 - nothing to do with currents. Likewise, `i1` is the input signal excited at port 1.

On the other hand, `o1,1` is the **o**utput signal measured at port 1 when port 1 is excited, `o2,1` is the output signal measured at port 2 when port 1 is excited, and so on.

In other words, "i" and "o" refer to the input and output time-domain signals, respectively. Not to be confused with the current symbol "_I_".

---

> Firstly if the MSL is ~50 Ohms, one port will be matched reasonable while the other is not at all.

I don't see how that's relevant, doesn't the [excitation signal](https://github.com/thliebig/openEMS/blob/782a7381bf8fb7e2100fa1358f076a1085d21bdd/FDTD/excitation.cpp#L168) (`Signal_volt[n] = cos(2.0*PI*f0*(t-9.0/(2.0*PI*fc)))*exp(-1*pow(2.0*PI*fc*t/3.0-3,2))`) describe the voltage being inputted/stimulated/fed into the port? In Python I use `port.ut_inc` to get that (`i1` in the plot), and that to me sounds like it should be completely irrelevant to the model or even the impedance of the port, or how much of the signal is reflected back into the port (I use `port.ut_ref` for that, and that gives me `o1,1` in the plot).

As an additional "counter-example" as to why I don't think it should be related to impedance mismatch but rather somehow related to the port impedance itself (for reasons I haven't figured out), I ran another example. Previously, the peak amplitude of port 2 (350 Ω)'s incident time-domain signal was greater than that of port 1 (50 Ω). One would think that, given the MSL impedance is ~50 Ω, the greater the deltaZ between port impedance and 50 Ω (and thus mismatch), the greater the incident time-domain amplitude.

But in this example (port 1 = 10 Ω, port 2 = 50 Ω), the opposite is observed:
<img width="1108" alt="Screenshot 2023-06-18 at 05 59 44" src="https://github.com/thliebig/openEMS-Project/assets/25392776/a585c4b3-b439-4daf-b66b-3c2e2cac6dec">

It seems that the more I increase the port impedance, the higher the peak amplitude of the incident time-domain signal gets:

| Port impedance | Peak incident voltage   |
|------------|-----------|
| 10 Ω       | ~510 μV   |
| 50 Ω       | ~550 μV   |
| 350 Ω      | ~1050 μV  |

I wonder if EC-FDTD somehow encodes the `Voltage = sqrt(Power * Resistance)` equation in there somewhere, and it explains what we're seeing. (maybe it's unrelated to EC - I'll try with a commercial non-EC solver tomorrow to see if the same thing is observed)

---

> Secondly if you have "mixed" impedance S-Parameter you have to be much more careful how to calculate them, just using voltage ratios is not it anymore...

Thanks for picking that up, indeed it should be `(o_i/i_j) * sqrt(Z_j)/sqrt(Z_i)` instead of `o_i/i_j`, where `o_i` is the measured signal at port `i` and `i_j` is the incident signal at port `j`:

```py
S_ij = (output_port.uf_ref / input_port.uf_inc) * sqrt(Z_input_port)/sqrt(Z_output_port)
```
Though I'm not sure if that suffices given the problem of different amplitudes above - perhaps I need to add a multiplication factor to keep both incident signals at the same amplitude?

**Reply by thliebig** [reply]
Well your input signals then seem to be the incoming voltages? Well the only point I want to get to is, that getting all the units right is important because we talk about physical properties everywhere...

I agree that the mismatch does not influence how to calculate the S-parameter, but it will strongly influence the amplitude and phase and everything regarding the incoming and reflected port voltages.
Because the ports just excite an electric field across their resistance and the time domain voltage and current across the port is recorded.  Transforming that into freq-domain and calculating incoming and reflected parts is purely mathematically. 
If you calculate the S-parameter taking all of the port impedences into account etc. everything should work out fine...
And all of this has nothing to do with the used numerical method... It's all a question of post-processing and understanding the recorded voltages and currents (which are of course all just recorded as signals over time)...

**Reply by 0xCoto** [reply]
> If you calculate the S-parameter taking all of the port impedences into account etc. everything should work out fine...

On a second thought, I don't think what I proposed (`S_ij = (output_port.uf_ref / input_port.uf_inc) * sqrt(Z_input_port)/sqrt(Z_output_port)`) would work, because if e.g. |S21| > 1 (or 0 dB) on a passive network like the example above, multiplying >1 by a positive value (since `sqrt(Z_input_port)/sqrt(Z_output_port)` will always be >0), will still keep |S21| > 1 (or 0 dB), which is clearly incorrect.

I feel like the right thing to do to correct this would be to excite a different amplitude to each port based on its impedance (not sure what the offset would have to be exactly), such that the excitation signals are identical. I tried checking the same thing with CST (port 1 = 350 Ω, port 2 = 50 Ω), and the input signals (`i1` and `i2`) are identical:

![port_1_350_ohm_vs_port_2_50_ohm_time](https://github.com/thliebig/openEMS-Project/assets/25392776/356e8e5f-85e3-4473-aa22-66d72b157668)

And the S-parameters appear correct (note: CST's port 1 is openEMS's port 2, and vice versa):

![port_1_350_ohm_vs_port_2_50_ohm_s_parameters](https://github.com/thliebig/openEMS-Project/assets/25392776/eed91cce-6597-4ace-a886-fb0095277f6b)

The reflection coefficients agree with openEMS (since the reference amplitude is the same as the excitation amplitude), but the transmission coefficients (S12, S21) are obviously different.

---

**Edit:** Taking a closer look at the initial time-domain screenshot, it seems that we're seeing more voltage at port 2 than the total voltage we fed into the system (from port 1), so something doesn't feel right...

![image](https://github.com/thliebig/openEMS-Project/assets/25392776/2b880883-7337-4502-9ff0-cb4b0654a9a9)

**Reply by thliebig** [reply]
I'm sorry but the labels on this curves do make no sense to me at all... what are voltages and what are not and what is incoming and what is reflected and what at which port and so on? Please use labels that make more sense if you want others to understand the curves.
As for higher voltage at port 2? If e.g. you have a current through your line then yes it will produce a higher voltage on port 2 than on port 1... 
You need to stop "I feel like the right thing to do" and readup on how S-parameter are defined and implement these correct quantities. If I remember correctly it is something like power in in port2 divided by power in port 1 and sqrt? But really you should check it...
But I repeat, everything is post-processing, the FDTD simulation is fully linear. Messing with the amplitudes there is unnecessary.

**Reply by VolkerMuehlhaus** [reply]
Regarding excitation voltages: 
don't we have a **current** source at the ports, in parallel to the port impedance? That is what I know from other FDTD implementations, and then higher port impedance obviouly generates a higher voltage for the incident pulse.

Regarding port 2 voltage being higher than incident voltage:
Reflection at a high impedance termination (350 Ohm) will lead to voltages that are higher than incident pulse. Voltage will double for an open end, and drop to zero for a short.

Regarding Sxy parameter calculation with mixed impedances: you need to be careful with that, it's not correct to just use voltage ratio like you do with equal port impedance models.

**Reply by thliebig** [reply]
You do not even need a reflection for higher voltages... Just imagine a MSL line than gets (slowly) thinner (higher impedance) along the line. The energy will travel happily and without any reflections along. But as the power stays (ideally) the same the voltage gets higher and the current gets lower...
Another good example is a transformer that transforms voltages up (or down) from one side (maybe feeding) to the other... 
I think the original formula (`S_ij = (output_port.uf_ref / input_port.uf_inc) * sqrt(Z_input_port)/sqrt(Z_output_port)`) looks correct, but I had no time to setup an example to check it.
Maybe someone can create a python example with a proper MSL transitioning e.g. from 50 to 100 Ohms?
I really think this thread shows that we need an example with mixed port impedance's...

**Comment by VolkerMuehlhaus** [comment]
> Maybe someone can create a python example with a proper MSL transitioning e.g. from 50 to 100 Ohms? I really think this thread shows that we need an example with mixed port impedance's...

I created such an example in Python, using 50 Ohm for port 1 and 100 Ohm for port 2. In between there is a 70 Ohm line that will do impedance transformation at quarter wavelength (around 840 MHz). File attached in ZIP format, or is it preferred to post all code in the editor here?

Equation S_ij = (output_port.uf_ref / input_port.uf_inc) * sqrt(Z_input_port)/sqrt(Z_output_port) is used here and works fine!

For comparison to another solver, I used Sonnet which can handle mixed port impedances. Results agree well, comparison is attached for reference. Enjoy!

[lambda_quarter_transformer_50Ohm_100Ohm.zip](https://github.com/thliebig/openEMS-Project/files/11824437/lambda_quarter_transformer_50Ohm_100Ohm.zip)

![quarterwave_line_mixed_impedance](https://github.com/thliebig/openEMS-Project/assets/47922981/cea334d8-5778-43bd-b48a-8052bdb8b6de)

![compare_sonnet_mixed_impedance](https://github.com/thliebig/openEMS-Project/assets/47922981/3fd9e2b0-8667-4876-8591-a16159aa9252)

**Reply by thliebig** [reply]
Thats looks and sounds very good. Thank you!

**Comment by 0xCoto** [comment]
> [...] will still keep |S21| > 1 (or 0 dB)

I made a miscalculation there - that was false. Thanks for testing @MWit83, I also ran a few more tests and the results agree with CST after applying the correction factor.
